invokeai/backend/model_manager/search.py

```diff
@@ -0,0 +1,195 @@
+# Copyright 2023, Lincoln D. Stein and the InvokeAI Team
+"""
+Abstract base class and implementation for recursive directory search for models.
+
+Example usage:
+```
+  from invokeai.backend.model_manager import ModelSearch, ModelProbe
+
+  def find_main_models(model: Path) -> bool:
+    info = ModelProbe.probe(model)
+    if info.model_type == 'main' and info.base_type == 'sd-1':
+        return True
+    else:
+        return False
+
+  search = ModelSearch(on_model_found=report_it)
+  found = search.search('/tmp/models')
+  print(found)   #  list of matching model paths
+  print(search.stats)  #  search stats
+```
+"""
+
+import os
+from abc import ABC, abstractmethod
+from pathlib import Path
+from typing import Callable, Optional, Set, Union
+
+from pydantic import BaseModel, Field
+
+from invokeai.backend.util.logging import InvokeAILogger
+
+default_logger = InvokeAILogger.get_logger()
+
+
+class SearchStats(BaseModel):
+    items_scanned: int = 0
+    models_found: int = 0
+    models_filtered: int = 0
+
+
+class ModelSearchBase(ABC, BaseModel):
```

Comment:
I wrote this some time ago, but I think at that point I was contemplating models being stored with their components in a normalized fashion so that there'd only be one instance of a shared VAE, for instance. Searching this structure would require a different implementation, so I wrote a base class for it.